components/emoji_picker/types/index.ts

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {EmojiCategory, Emoji} from 'mattermost-redux/types/emojis';
+
+import {
+    CATEGORY_HEADER_ROW,
+    EMOJIS_ROW,
+    NAVIGATE_TO_NEXT_EMOJI,
+    NAVIGATE_TO_NEXT_EMOJI_ROW,
+    NAVIGATE_TO_PREVIOUS_EMOJI,
+    NAVIGATE_TO_PREVIOUS_EMOJI_ROW,
+} from 'components/emoji_picker/constants';
+
+export type Category = {
+    className: string;
+    emojiIds?: string[];
+    id: string;
+    message: string;
+    name: EmojiCategory;
+};
+
+export type Categories = Record<EmojiCategory, Category>;
+
+export type CategoryOrEmojiRow<
```

Comment:
I think this might be better as a union type. It should clean stuff up if we define it like below and add some type predicates
```typescript
export type CategoryOrEmojiRow = CategoryHeaderRow | EmojiRow;

export type CategoryHeaderRow = {
    index: number;
    type: 'category_header_row';
    categoryName: string;
}

export type EmojiRow = {
    index: number;
    type: 'emoji_row';
    // ...
}
```

For the type predicates, we can just have them look at the value of `type` to determine what the type of a row is. We might even be able to skip defining the predicates and just check the value of `type` directly as well like

```typescript
let row = ...; // TS sees row as a CategoryOrEmojiRow

if (row.type === 'category_header_row') { // Or use a type predicate
    // TS sees row as a CategoryHeaderRow
} else {
    // TS sees row as an EmojiRow
}
```

Reply:
✅  Updated as per request/discussion